### PR TITLE
[UI Tests] - Add step on Buildkite to run Jetpack UI Tests

### PR DESCRIPTION
**Pull request by jostnes**
### Description:
This PR adds the Jetpack UI tests to run on each build. It's set up to run as a separate step:
image
image

While @tiagomar and I discussed making the Jetpack tests optional (in the interest of time) looks like both tests run in parallel on Firebase and complete around the same time, so build time is not affected by this change. Because of this, I think it's okay not to make it optional. Please let me know if there are other thoughts about this.

Also included in the PR (these are not in separate PR as I found about these issues while trying to run the tests in CI 🙏 ):
1. https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/5b0366e59d45ae760b78470d571c3243225cd49e fix by @irfano to fix a hilt error that was crashing the app at the start of test (only happening on CI)
2. UI test fixes as the Logout button is not displayed when running on Pixel 2 (the device used on Firebase), which caused failures when trying to log out: 
image

### Testing:
Ensure that both WordPress and Jetpack tests are passing in CI and locally.

### CODE REVIEW

File: fastlane/lanes/test.rb
Comment (jostnes) [inline fastlane/lanes/test.rb]: For Jetpack, only targetting UI tests under the `org.wordpress.android.e2e` package

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java
Comment (jostnes) [inline WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java:58]: Updated to wait for a menu item that's listed near the top of the menu instead of the last item.

**Comment by pachlava** [comment]
> Please let me know if there are other thoughts about this.

@jostnes 👋 

I'm curious if disabling the WP ones on CI was considered an option? Even though they run in parallel now, this might not always be the case, this probably depends on the amount of free workers 🤔. Alternatively, they can be made optional, instead of the JP ones?

File: fastlane/lanes/test.rb
Comment (ParaskP7) [reply fastlane/lanes/test.rb]: 👋 @jostnes !

Why is that so, why are we targeting only those [15 test cases](https://console.firebase.google.com/project/api-project-108380595987/testlab/histories/bh.3ce77db975c1f0c9/matrices/8732295795092882555/executions/bs.6507729498e793b9/test-cases) for Jetpack, with one of these skipped anyway? 🤔 

PS: For WordPress we are targeting [121 test cases](https://console.firebase.google.com/project/api-project-108380595987/testlab/histories/bh.b3d1e88d8b09c8be/matrices/8018251029287387893/executions/bs.bfc10ce437d81849/test-cases) overall, with the same one being skipped (just like for Jetpack above).

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (pachlava) [inline WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: I'm curious about the reason for this removal 🤔 It does save time (theoretically, 0.1 sec, in case if the element is already on screen), but we are no longer protected from the slow screen load.

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java
Comment (ParaskP7) [reply WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java:58]: 👋 @jostnes !

I understand that this fixes UI tests as the `Logout` button is not displayed when running on Pixel 2 (the device used on Firebase), which caused failures when trying to log out. But why is that happening now, and becomes a requirement for this PR, I guess this is just `Jetpack` UI test related, right? 🤔 

Also, as part of this change you removed 3 of `waitForElementToBeDisplayed(...)`, 2 from `WPSupportUtils.scrollToThenClickOn(...)` and 1 from `MePage.logout()`. Why was that a requirement for this PR? 🤔

File: fastlane/lanes/test.rb
Comment (ParaskP7) [inline fastlane/lanes/test.rb]: Suggestion (💡): This lane is an almost duplicate of the `build_and_run_wordpress_instrumented_test` above. The only diffs I am seeing are the below:
- lane:
    - `assembleWordPressVanillaDebug` and `assembleJetpackVanillaDebug` -> `WordPress` and `Jetpack`)
    - `assembleWordPressVanillaDebugAndroidTest` and `assembleJetpackVanillaDebugAndroidTest`  -> `WordPress` and `Jetpack`
- test_apk_path:
    - `wordpressVanilla` and `jetpackVanilla` -> `wordpress` and `jetpack`
    - `org.wordpress.android-wordpress-vanilla-debug-androidTest.apk` and `org.wordpress.android-jetpack-vanilla-debug-androidTest.apk` -> `wordpress` and `jetpack`
- apk_path:
    - `wordpressVanilla` and `jetpackVanilla` -> `wordpress` and `jetpack`
    - `org.wordpress.android-wordpress-vanilla-debug.apk` and `org.wordpress.android-jetpack-vanilla-debug.apk` -> `wordpress` and `jetpack`
- test_targets:
    - `notPackage org.wordpress.android.ui.screenshots` and `package org.wordpress.android.e2e` -> 👍
- annotation_ctx:
    - `firebase-test-wordpress-vanilla-debug` and `firebase-test-jetpack-vanilla-debug` -> `wordpress` and `jetpack`

From the above you extract the following pattern for de-duplication needs:
- `app` key for the `WordPress` and `Jetpack` values.
- Another, or the same `app` key, for the `wordPress` and `jetpack` values.
- `test_targets` key for its corresponding values.

I think it is important that we remove this duplication to avoid future problems when one of the lanes is updated without the other one being updated too, and this way, we'll also avoid maintaining more than we should.

As such, I suggest creating one parent lane (`build_and_run_instrumented_test`) which would hold parameters like `app` and `test_targets` and accordingly pass them to the children lanes (`build_and_run_wordpress_instrumented_test` and `build_and_run_jetpack_instrumented_test`).

PS: On the `app` key, I think we already have something like that in place. Maybe @AliSoftware can add some ideas on how to make my above de-duplication suggest more robust.

**Review by ParaskP7 — COMMENTED** [review]
👋 @jostnes !

I have reviewed and tested this PR as per the description, everything LGTM, good job there Jos! 🌟 

FYI: I do have some suggestions (💡) and question (❓) for you to consider. Thus, I am not going to approve this PR just yet, but also, I am not going to block it, I'll just add my comments and let you take it from there. 🙏 

-----

Suggestion (💡): While reviewing this PR, I noticed the below failures on both the [WP Instrumented tests](https://buildkite.com/automattic/wordpress-android/builds/9455#0186739f-05ec-42ff-97b3-eb819fd14997) and the [JP Instrumented tests](https://buildkite.com/automattic/wordpress-android/builds/9455#0186739f-05f0-402b-9e5c-ad8f0c7aaef6) CI jobs.


```
- WP:
  2023-02-21 11:31:49 FATAL  Failed to remove annotation:
  DELETE https://agent.buildkite.com/v3/jobs/0186739f-05ec-42ff-97b3-eb819fd14997/annotations/firebase-test-wordpress-vanilla-debug: 410 Annotation no longer exists
- JP:
  2023-02-21 11:30:50 FATAL  Failed to remove annotation:
  DELETE https://agent.buildkite.com/v3/jobs/0186739f-05f0-402b-9e5c-ad8f0c7aaef6/annotations/firebase-test-jetpack-vanilla-debug: 410 Annotation no longer exists
```

To be honest, even before noticing those, I wanted to click on the Buildkite annotation solution for Firebase Test Lab that @AliSoftware created, only to realize/remember that (by design) this annotation was removed on success (8e2c00a76e4ac170a961fd33cbbdf1e8e952fcb1).

That made me wonder whether we could benefit from having this Buildkite annotation solution for Firebase Test Lab even when everything is successful, especially now that we have 2 of those `Instrumented tests` running in parallel.

Wdyt about both the failures further above and my suggestion above having those annotations on success @jostnes and @AliSoftware ? 🤔 🤔

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (ParaskP7) [reply WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: I asked the same [here](https://github.com/wordpress-mobile/WordPress-Android/pull/17950#discussion_r1115586798), @jostnes feel free to respond to @pachlava here only in order to avoid responding in multiple places. 👍

File: fastlane/lanes/test.rb
Comment (AliSoftware) [reply fastlane/lanes/test.rb]: +1 on [DRYing](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) the code.

As for tips, you could use the same technique as what we use [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L137-L139). Basically:
 - [We have a Dictionary here which is defining a set of constants with their different values for WordPress vs Jetpack](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/Fastfile#L3-L24)
 - On our lanes that are in the same case as yours, i.e. we do mostly the same for WordPress vs Jetpack with only a small set of variations, we use our [`app = get_app_name_option!(options)` helper method](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/release.rb#L314-L321) to read the `app:` parameter that is passed to the lane (and normalize it and error nicely if it's missing), and then we use `APP_SPECIFIC_VALUES[app][:some_key]` to read the app-specific value for a given `some_key` constant defined from this dictionary

This allows you to avoid having to pass many parameters to the lane (all having to be consistent with one another) to invoke things for WordPress vs Jetpack, or having to create a parent lane accepting many parameters then add two children lanes invoking it with the right values (like Petros suggested and like we used to do a while ago before we started to use this `APP_SPECIFIC_VALUES` technique)… and instead you [just have to pass `app:wordpress` or `app:jetpack` when invoking those kind of lanes](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L132-L134) and the implementation of your lane will pick the constants for the right app from the `APP_SPECIFIC_VALUES` constants, ensuring consistency in the different values being used for a given `app:`.

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (jostnes) [reply WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: This piece of code actually did not work with the wait here 😅. However, we weren't aware of the bug as where it was used previously, all required no scrolling (when running WP tests, the Logout button was already on screen so scrolling wasn't required). 

At this `wait`, the test was waiting for an element that wasn't visible on the screen yet to be visible, a condition that it would not meet, so the test would fail. After removing the `wait`, the `scroll` works (as you can see, all `waits` that were removed are before a `scroll`). After testing multiple times (locally) and once in CI, it looks like the wait is not necessary, so I think it would be better to remove it than keep it (in this case, the position would move to after the scroll) when it doesn't serve any real purpose.

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java
Comment (jostnes) [reply WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java:58]: Replied with more details in the other comment to summarize, the menu items on WordPress and Jetpack are different, which I only found out from CI testing in this PR.

File: fastlane/lanes/test.rb
Comment (jostnes) [reply fastlane/lanes/test.rb]: Thanks for the suggestions! I'm not familiar with ruby and wasn't sure how to make the code better, so the tips are very helpful. I'll update this.

File: fastlane/lanes/test.rb
Comment (jostnes) [reply fastlane/lanes/test.rb]: Only the 15 test cases are UI tests (e.g. require the Jetpack app to be installed and tested on). The other tests look like integration or API tests. Since the codebase is shared and there are no visual differences between the two, I don't think there's value in including those tests, at least in this case.

Please let me know if you think otherwise, and we can discuss it.

**Comment by jostnes** [comment]
> I'm curious if disabling the WP ones on CI was considered an option? 

We didn't consider that (we only considered the opposite), in the long run, the tests running on WP and JP would start to differ as the apps start adding/removing features, so it would also be good to have at least some form of UI tests done on both. 

> Even though they run in parallel now, this might not always be the case, this probably depends on the amount of free workers 🤔. Alternatively, they can be made optional, instead of the JP ones?

Is your main concern about having them run in parallel because of the workers availability? (i.e. it would block other builds from running/completing or end up taking a longer time to build when they don't run in parallel?) 

Shamelessly pinging @ParaskP7 / @AliSoftware to see if you have thoughts on this. I'm not too familiar with Buildkite and the availability of the workers, do you foresee running both WP and JP instrumented tests in the same build going to be an issue?

File: fastlane/lanes/test.rb
Comment (jostnes) [reply fastlane/lanes/test.rb]: Simplified code in these commits: https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/5b9a9fc827740a1cf89fdfeae62ac9b0529b5d5b and https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/ea09f0407c8ddd1d38b4e5f91d8023a71b7b01ce (I missed one change in the first commit 😅)

I decided to use the `APP_SPECIFIC_VALUES` way that @AliSoftware suggested so there is only 1 lane that should now work for both apps.

File: fastlane/Fastfile
Comment (AliSoftware) [inline fastlane/Fastfile]: 🔍 Nitpick: Given those constants are declared in the global space, I'd suggest to name that key something more like `:uitest_targets` or `instrumented_test_target`. That way if later we need a similar per-app key/constant but for _Unit_ Tests, we'd not risk confusion between the two.

Comment (jostnes) [reply fastlane/Fastfile]: Yup that makes sense, updated in https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/26e76361a98898d2e082748fe96e0e8f60ec8366

**Comment by AliSoftware** [comment]
> > Even though they run in parallel now, this might not always be the case, this probably depends on the amount of free workers 🤔. Alternatively, they can be made optional, instead of the JP ones?
> 
> Is your main concern about having them run in parallel because of the workers availability? (i.e. it would block other builds from running/completing or end up taking a longer time to build when they don't run in parallel?)
> 
> Shamelessly pinging @ParaskP7 / @AliSoftware to see if you have thoughts on this. I'm not too familiar with Buildkite and the availability of the workers, do you foresee running both WP and JP instrumented tests in the same build going to be an issue?

That's a good question… which doesn't have an easy answer 😅  It's all a matter of balance to find.

For the case a hand, I'd say that, given:
 - Those UI Test suites "only" take ≈27mn each
 - They run on AMIs, so no hard limit in total number of workers (≈auto-scaling)—unlike with our Mac fleet of physical machines—and thus less chance that they wouldn't be able to be run in parallel
 - They are only run if a PR is open (and not run if one pushes a branch but haven't got a PR for that branch opened yet)

Then I think we can just keep things how you've done it here, running both UI Test Suites for WordPress and Jetpack, as they are likely to always run in parallel without worker count limits. (There might be a question of cost impact though, but I'm pretty sure the current policy from Jeremy and NB is "if we need to pay for more AMI worker time we're ready to pay rather than trying to reduce test suites just to cut costs, because testing and quality is important").

💭  That might be a different situation for iOS (assuming you plan to replicate a similar setup to this PR for iOS soon?), because for Mac builds we have a limited fleet of physical workers (even if we're currently discussing with our provider to rent and provision more Macs in the near future), so more risk that they would not be run in parallel if not enough Mac workers available to do so. It might also depend on if those iOS UI Tests end up taking way longer than the Android UI Test suites you have here. But we can rediscuss that when it comes to it.

File: fastlane/lanes/test.rb
Comment (pachlava) [reply fastlane/lanes/test.rb]: My two cents is that these additional 106 instrumentation tests run for only 13 seconds.

> Since the codebase is shared and there are no visual differences between the two, I don't think there's value in including those tests, at least in this case.

The same logic can be applied to UI tests, then 🤔 I see value in running them all, especially since they are fast, _and_ we save ourselves from accidentally forgetting to include them back again later, when there are more differences between the apps.

**Comment by pachlava** [comment]
> Is your main concern about having them run in parallel because of the workers availability? (i.e. it would block other builds from running/completing or end up taking a longer time to build when they don't run in parallel?)

It was one of the main concerns, yes, and @AliSoftware gave a perfect answer (thank you!) 👍

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (pachlava) [reply WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: Thank you for the explanation @jostnes! I agree that if we can shorten something, and still have the tests work in the long run the same way, it's justified. My concern was that it was a part of the PR that does something different. However, these words of yours:

> which I only found out from CI testing in this PR.

make it clear 👍

File: fastlane/lanes/test.rb
Comment (ParaskP7) [reply fastlane/lanes/test.rb]: Awesome update @jostnes , Mr. @AliSoftware you are the best! 🌟 🌟 🌟

File: fastlane/Fastfile
Comment (ParaskP7) [inline fastlane/Fastfile]: Minor (🔍): Consider removing this extra empty line @jostnes . 🙏

File: fastlane/lanes/test.rb
Comment (ParaskP7) [reply fastlane/lanes/test.rb]: @jostnes I do agree with @pachlava on that. 💯 

Actually, even if they would have caused minutes of increased test run, I think it is better to start with the full suite of tests to be run on both, the `Jetpack` and `WordPress` apps, and maybe then, take it from there, if need be, when features start getting removed from `WordPress`, and thus try to separate those somehow to gain in performance of those tests completing.

For now, I would go with consistency just so that we can verify that every test that can be run on `Jetpack` is able to run and is as operational as possible.

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (ParaskP7) [reply WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: Same, thanks for the explantation @jostnes , it is a 👍 from my side too! 🙇 

Suggestion (💡): Might I suggest something that I usually follow myself (for next time). In order for me to avoid explaining something like that here and there, on my PRs or otherwise, I usually prefer to add that to my commit, as part of the its description. This way, the reviewer, or any other engineer that would check this commit/change later on would know why that change happened in full detail, like reading a diary of the author's mind at that point, without needing to have to dig more into that or even ask such questions to the author. I hope that makes sense. This is because a commit like `remove unnecessary wait` cannot help much in these situations, even if you add a PR description like `UI test fixes as the Logout button is not displayed...`.

Btw, sometimes I post these questions of mine just to make the author reply on them and have them re-think it, even re-test it, so they can try and explain it to someone else afterwards and see if that will trigger a change. Then, if the author can explain it, no matter if I agree, of if I can understand the change (🤣),  or even if I don't agree, this will make the author accountable for it and me comfortable to keep things going forward, thus me giving the 👍 on merging, etc.

PS: Please don't mind my rumbling on the above too much, I am just trying to help anyway I can and make us more efficient in terms of commits/code reviews, but I can always be wrong, so please feel free to debate on anything you feel that doesn't click with you, on anything I say! 🙏

**Comment by ParaskP7** [comment]
> It was one of the main concerns, yes, and @AliSoftware gave a perfect answer (thank you!) 👍

:this: 💯 :this: Thanks for this detailed answer @AliSoftware ! 🥇

File: fastlane/lanes/test.rb
Comment (irfano) [reply fastlane/lanes/test.rb]: I've checked those tests outside of the `e2e` package.
- [URITest](https://github.com/wordpress-mobile/WordPress-Android/blob/ed574ea833797f71e28fb2abd060fee7d305376c/WordPress/src/androidTest/java/URITest.java) and [URLTest](https://github.com/wordpress-mobile/WordPress-Android/blob/ed574ea833797f71e28fb2abd060fee7d305376c/WordPress/src/androidTest/java/URLTest.java) are testing `java.net` classes, which doesn't make sense. I think we can remove these. They include **10** tests. This can be another issue.
- Other tests are testing features identical for WP and JP. But still, I believe we should test all for WP and JP. 
- On [UI tests](https://github.com/wordpress-mobile/WordPress-Android/tree/trunk/WordPress/src/androidTest/java/org/wordpress/android/e2e) document, it says `2. Run the tests in org.wordpress.android.e2e`. I think we should move all tests under `e2e` package, or update the document. This is just an idea for another issue.

File: WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java
Comment (jostnes) [reply WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:101]: Thank you for the suggestion @ParaskP7! I guess the change was clear to me and I incorrectly assumed that it would be clear for others as well, I'll be more mindful of this in the future and will include as much information for next time! 🙇

File: fastlane/lanes/test.rb
Comment (jostnes) [reply fastlane/lanes/test.rb]: Thank you everyone for the input! Since there are opinions around this and I have some unanswered questions, I make the update to include the tests to be on both and WP and JP steps, change made in this commit https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/037a16df4229980fd6bb0f300c4b0dd466f6093e. 

For points 1 and 3 that you mentioned @irfano are out of scope of this PR so I created this [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/18014) so we don't forget about it.

I'd like to share my perspective about my decision to only include only the UI tests on Jetpack in case it's not clear. To be honest duration/time taken was not the main factor I considered, I was thinking more in the line of necessity. I checked out a sample of the non-UI connected tests (e.g. [testSafeToAddAuthToken1](https://github.com/wordpress-mobile/WordPress-Android/blob/8c610143de99b5dbacacf4e2b1177f3e3fe91d41/WordPress/src/androidTest/java/org/wordpress/android/util/WPUrlUtilsTest.java#L29) and [testGetHost1](https://github.com/wordpress-mobile/WordPress-Android/blob/8c610143de99b5dbacacf4e2b1177f3e3fe91d41/WordPress/src/androidTest/java/org/wordpress/android/util/UrlUtilsTest.java#L12)) and from the samples, the tests don't seem to run on an emulator (making it unaware if the app is WP or JP, which is the main difference between the two steps). Though there are questions that I still don't know the answers to:

1. I wonder if the non UI connected tests will diverge between WP and JP like how the UI tests will be in the future as we add/remove features to the apps?
2. I'm not sure if running it on Jetpack will have additional value or will it be a redundant test run (will both tests always end with the same exact results and what are the differences between it running on WP and JP steps since it doesn't run on the app)?

Since there are unanswered questions and making an assumption is probably not a good way to make this decision, making it run on both for now seems like the way to go. We can change it later if needed once we know for certain the answers to these.

Thank you again all for the input! 🙇

**Comment by jostnes** [comment]
👋 @pachlava @ParaskP7 thanks for taking the time to review this PR! I believe the issues have been addressed, can you take a second look? I'd like to have this merged soon if possible.

cc: @tiagomar

**Comment by ParaskP7** [comment]
👋 @jostnes !

> 👋 @pachlava @ParaskP7 thanks for taking the time to review this PR! I believe the issues have been addressed, can you take a second look? I'd like to have this merged soon if possible.

Thank you so much for dealing with all our comments and applying changes based on them. From my side it is a 👍 and I am approving. Feel free to wait on @pachlava and/or others to approve before merging, or merging this PR asap to get you unblocked.

**Review by ParaskP7 — APPROVED** [review]

**Review by pachlava — APPROVED** [review]
> 👋 @pachlava @ParaskP7 thanks for taking the time to review this PR! I believe the issues have been addressed, can you take a second look? I'd like to have this merged soon if possible.

Thank you for the further explanations and adjustments @jostnes! It's good to go from my side as well :shipit: 👍

**Pull request merged**
